src/pyhf/optimize/common.py

```diff
@@ -0,0 +1,129 @@
+"""Common Backend Shim to prepare minimization for optimizer."""
+from .. import get_backend, default_backend
+from ..tensor.common import _TensorViewer
+
+
+def _get_tensor_shim():
+    """
+    A shim-retriever to lazy-retrieve the necessary shims as needed.
+
+    Because pyhf.tensor is a lazy-retriever for the backends, we can be sure
+    that tensorlib is imported correctly.
+    """
+    tensorlib, _ = get_backend()
+    if tensorlib.name == 'numpy':
+        from .opt_numpy import wrap_objective as numpy_shim
+
+        return numpy_shim
+
+    if tensorlib.name == 'tensorflow':
+        from .opt_tflow import wrap_objective as tflow_shim
+
+        return tflow_shim
+
+    if tensorlib.name == 'pytorch':
+        from .opt_pytorch import wrap_objective as pytorch_shim
+
+        return pytorch_shim
+
+    if tensorlib.name == 'jax':
+        from .opt_jax import wrap_objective as jax_shim
+
+        return jax_shim
+    raise ValueError(f'No optimizer shim for {tensorlib.name}.')
+
+
+def shim(
+    objective,
+    data,
+    pdf,
+    init_pars,
+    par_bounds,
+    fixed_vals=None,
+    do_grad=False,
+    do_stitch=False,
+):
+    """
+    Prepare Minimization for Optimizer.
+
+    Args:
+        objective (`func`): objective function
+        data (`list`): observed data
+        pdf (`pyhf.pdf.Model`): model
+        init_pars (`list`): initial parameters
+        par_bounds (`list`): parameter boundaries
+        fixed_vals (`list`): fixed parameter values
+
+    .. note::
+
+        ``minimizer_kwargs`` is a dictionary containing
+
+          - ``func`` (`func`): backend-wrapped ``objective`` function
+          - ``x0`` (`list`):  modified initializations for minimizer
+          - ``jac`` (`func`): callable that accepts same parameters as the input ``objective`` but returns the gradient
+          - ``bounds`` (`list`): modified bounds for minimizer
+          - ``fixed_vals`` (`list`): modified fixed values for minimizer
+
+    .. note::
+
+        ``stitch_pars(pars, stitch_with=None)`` is a callable that will
+        stitch the fixed parameters of the minimization back into the unfixed
+        parameters.
+
+    .. note::
+
+        ``do_stitch`` will modify the ``init_pars``, ``par_bounds``, and ``fixed_vals`` by stripping away the entries associated with fixed parameters. The parameters can be stitched back in via ``stitch_pars``.
+
+    Returns:
+        minimizer_kwargs (`dict`): arguments to pass to a minimizer following the ``scipy.optimize.minimize`` API (see notes)
+        stitch_pars (`func`): callable that stitches fixed parameters into the unfixed parameters
+    """
+    tensorlib, _ = get_backend()
+
+    if do_stitch:
+        all_init = default_backend.astensor(init_pars)
+        all_idx = default_backend.astensor(range(pdf.config.npars), dtype='int')
+
```

Comment:
why do these use default_backend and not `tensorlib`?

Reply:
we don't do any computations -- just slicing. Those two variables you highlighted go into here

```python
        variable_idx = [x for x in all_idx if x not in fixed_idx]
        variable_init = default_backend.tolist(all_init[variable_idx])
```

`variable_init` is passed back out, while `variable_idx` is passed into `TensorViewer`. We could use `tensorlib` here entirely and drop `default_backend`.

Reply:
fixed it up.